### PR TITLE
Add heal elements into client chain

### PR DESCRIPTION
**Pull request by d-uzlov**
## Issue link

fixes https://github.com/networkservicemesh/sdk/issues/938

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionality
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/chains/client/client.go [hunk 1]
```diff
@@ -48,32 +52,74 @@ type Option func(c *clientOptions)
...
+		)...,
+	)
+	return rv
+}
+
+// NewServerSideClient returns an NSM client for (2.) and (3.) cases.
+// This function is intended to be a part of a client factory, and therefore should never ba called directly.
+//  ctx - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
+//  cc  - grpc.ClientConnInterface for the endpoint to which this client should connect.
+func NewServerSideClient(ctx context.Context, cc grpc.ClientConnInterface, clientOpts ...Option) networkservice.NetworkServiceClient {
```
Comment (edwarnicke) [inline pkg/networkservice/chains/client/client.go:122]: I am deeply uncertain that introducing a function like this is a good idea to solve our problem

**Review by edwarnicke — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/client/client.go (hunk 1 above)
Comment (d-uzlov) [reply pkg/networkservice/chains/client/client.go:122]: Could you elaborate on your concerns, please?

File: pkg/networkservice/chains/client/client.go [hunk 2]
```diff
@@ -97,24 +143,19 @@ func NewClient(ctx context.Context, cc grpc.ClientConnInterface, clientOpts ...O
 	return rv
 }
 
-// Factory creates a networkservice.NetworkServiceClient by passed context.Cotnext and grpc.ClientConnInterface
-type Factory = func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient
-
 // NewCrossConnectClientFactory - returns a (2.) case func(cc grpc.ClientConnInterface) NSM client factory.
-func NewCrossConnectClientFactory(clientOpts ...Option) Factory {
+func NewCrossConnectClientFactory(clientOpts ...Option) connect.ClientFactory {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/client/client.go:147]: I thought we'd removed this.
```suggestion
```

File: pkg/networkservice/chains/client/client.go (hunk 1 above)
Comment (denis-tingaikin) [reply pkg/networkservice/chains/client/client.go:122]: I agreed with @edwarnicke 

@d-uzlov Can we just inline this into NewClientFactory

File: pkg/networkservice/common/connect/server.go [hunk 3]
```diff
@@ -25,22 +25,25 @@ import (
...
 	"github.com/networkservicemesh/sdk/pkg/tools/multiexecutor"
 )
 
+// ClientFactory is used to created new clients when new connection is created.
+type ClientFactory = func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient
+
 type connectServer struct {
 	ctx               context.Context
-	clientFactory     client.Factory
+	clientFactory     ClientFactory
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/connect/server.go:46]: Why is this changed?

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/common/connect/server.go (hunk 3 above)
Comment (d-uzlov) [reply pkg/networkservice/common/connect/server.go:46]: `connect` package depend on `chains`/`client` package (but only takes `client.Factory` definition from it), and now that `NewClient` uses `connect` server, we got a dependency loop.

File: pkg/networkservice/chains/client/client.go [hunk 4]
```diff
@@ -48,32 +52,74 @@ type Option func(c *clientOptions)
...
+		opt(opts)
+	}
+	rv = chain.NewNetworkServiceClient(
+		append(
+			append([]networkservice.NetworkServiceClient{
+				updatepath.NewClient(opts.name),
+				serialize.NewClient(),
+				refresh.NewClient(ctx),
+				metadata.NewClient(),
+			}, opts.additionalFunctionality...),
```
Comment (Bolodya1997) [inline pkg/networkservice/chains/client/client.go:95]: This should be moved into the connect client chain, because `sendfd` wouldn't simply work in the other case.

**Review by denis-tingaikin — APPROVED** [review]
@edwarnicke Could you also have a look?

**Comment by denis-tingaikin** [comment]
Waits for https://github.com/networkservicemesh/sdk/pull/927

File: pkg/networkservice/chains/client/client.go [hunk 5]
```diff
@@ -73,7 +74,7 @@ func WithAuthorizeClient(authorizeClient networkservice.NetworkServiceClient) Op
 // NewClient - returns a (1.) case NSM client.
 //             - ctx    - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
 //             - cc - grpc.ClientConnInterface for the endpoint to which this client should connect
-func NewClient(ctx context.Context, cc grpc.ClientConnInterface, clientOpts ...Option) networkservice.NetworkServiceClient {
+func NewClient(ctx context.Context, connectTo *url.URL, dialOptions []grpc.DialOption, clientOpts ...Option) networkservice.NetworkServiceClient {
```
Comment (Bolodya1997) [inline pkg/networkservice/chains/client/client.go:77]: Probably it would be better to add new option instead of passing here a slice.
```go
func WithDialOptions(dialOptions ...grpc.DialOption) Option {...}
```

File: pkg/networkservice/common/connect/networkservice_client.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/heal"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
+)
+
+// NewNetworkServiceClient - adapts connect server chain to client
+func NewNetworkServiceClient(ctx context.Context, connectTo *url.URL, dialOptions []grpc.DialOption, additionalFunctionality ...networkservice.NetworkServiceClient) networkservice.NetworkServiceClient {
```
Comment (Bolodya1997) [inline pkg/networkservice/common/connect/networkservice_client.go:34]: So I guess here we also can introduce option pattern for dial options, additional functionality as in registry connect:
https://github.com/networkservicemesh/sdk/blob/main/pkg/registry/common/connect/option.go
https://github.com/networkservicemesh/sdk/blob/main/pkg/registry/common/connect/nse_client.go#L46

File: pkg/networkservice/chains/client/client.go (hunk 5 above)
Comment (Bolodya1997) [reply pkg/networkservice/chains/client/client.go:77]: @glazychev-art
Not sure if you are receiving notifications for this, so calling you :)

File: pkg/networkservice/chains/client/client_heal_test.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	select {
+	case err := <-endpoint.Serve(serverCtx, serverURL, nse):
+		require.NoError(t, err)
+	default:
+	}
+
+	return serverCancel
+}
+
+func checkURLFree(u string) func() bool {
```
Comment (Bolodya1997) [inline pkg/networkservice/chains/client/client_heal_test.go:80]: We have such check in `grpcutils` now, please use it from there.

File: pkg/networkservice/chains/client/client.go [hunk 8]
```diff
@@ -70,51 +74,73 @@ func WithAuthorizeClient(authorizeClient networkservice.NetworkServiceClient) Op
...
+		for _, opt := range clientOpts {
+			opt(opts)
+		}
+		rv = chain.NewNetworkServiceClient(
+			append(
+				append([]networkservice.NetworkServiceClient{
+					updatepath.NewClient(opts.name),
+					serialize.NewClient(),
+					heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
+					refresh.NewClient(ctx),
```
Comment (Bolodya1997) [inline pkg/networkservice/chains/client/client.go:138]: It would be better to use first `refresh` and then `heal` chain elements:
```go
...
refresh.NewClient(ctx),
heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
...
```

File: pkg/networkservice/chains/client/client.go [hunk 9]
```diff
@@ -70,51 +74,73 @@ func WithAuthorizeClient(authorizeClient networkservice.NetworkServiceClient) Op
...
+			updatepath.NewClient(opts.name),
+			serialize.NewClient(),
+			refresh.NewClient(ctx),
+			metadata.NewClient(),
+		},
 			opts.authorizeClient,
-			networkservice.NewNetworkServiceClient(cc),
+			connect.NewNetworkServiceClient(ctx, connectTo,
+				connect.WithDialOptions(opts.dialOptions...),
+				connect.WithAdditionalFunctionality(opts.additionalFunctionality...),
```
Comment (Bolodya1997) [inline pkg/networkservice/chains/client/client.go:114]: Please add here `heal` client after the `additionalFunctionality` :)

File: pkg/networkservice/chains/client/client.go (hunk 8 above)
Comment (Bolodya1997) [reply pkg/networkservice/chains/client/client.go:138]: Probably even:
```go
...
refresh.NewClient(ctx),
metadata.NewClient(),
heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
...
```

File: pkg/networkservice/common/connect/networkservice_client.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		chain.NewNetworkServiceServer(
+			clienturl.NewServer(connectTo),
+			heal.NewServer(ctx, nil),
+			NewServer(
+				ctx,
+				func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
+					return chain.NewNetworkServiceClient(
+						append(
+							opts.clientAdditionalFunctionality,
+							heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
```
Comment (Bolodya1997) [inline pkg/networkservice/common/connect/networkservice_client.go:49]: Please remove `heal` from here, it should come as a part of additional functionality.

File: pkg/networkservice/chains/client/client.go (hunk 5 above)
Comment (glazychev-art) [reply pkg/networkservice/chains/client/client.go]: Done

File: pkg/networkservice/common/connect/networkservice_client.go (hunk 6 above)
Comment (glazychev-art) [reply pkg/networkservice/common/connect/networkservice_client.go]: Done

File: pkg/networkservice/chains/client/client_heal_test.go (hunk 7 above)
Comment (glazychev-art) [reply pkg/networkservice/chains/client/client_heal_test.go]: Done

File: pkg/networkservice/chains/client/client.go (hunk 8 above)
Comment (glazychev-art) [reply pkg/networkservice/chains/client/client.go:137]: Done

File: pkg/networkservice/common/connect/networkservice_client.go (hunk 10 above)
Comment (glazychev-art) [reply pkg/networkservice/common/connect/networkservice_client.go:49]: But we need `cc` for heal.NewClient(...). Probably we need to use extra option `WithHeal`?

Comment (Bolodya1997) [reply pkg/networkservice/common/connect/networkservice_client.go:49]: OK, I see. So probably it would be better to use here `ClientFactory` instead of `additionalFunctionality` if so. `connect` shouldn't now anything about `heal`.

**Review by Bolodya1997 — APPROVED** [review]

File: pkg/networkservice/common/connect/networkservice_client.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	clientOpts ...Option,
+) networkservice.NetworkServiceClient {
+	var opts = &connectServer{}
+	for _, opt := range clientOpts {
+		opt(opts)
+	}
+	return adapters.NewServerToClient(
+		chain.NewNetworkServiceServer(
+			clienturl.NewServer(connectTo),
+			heal.NewServer(ctx, nil),
```
Comment (Bolodya1997) [inline pkg/networkservice/common/connect/networkservice_client.go:47]: This is totally incorrect:
1. `connect` -> `heal` dependency
2. heal server should take a pointer to the whole chain as a 2 parameter

**Review by Bolodya1997 — CHANGES_REQUESTED** [review]

File: pkg/networkservice/common/connect/networkservice_client.go (hunk 11 above)
Comment (Bolodya1997) [reply pkg/networkservice/common/connect/networkservice_client.go:47]:
```go
rv := new(networkservice.NetworkServiceClient)
	*rv = chain.NewNetworkServiceClient(
		updatepath.NewClient(opts.name),
		serialize.NewClient(),
		refresh.NewClient(ctx),
		metadata.NewClient(),
		adapters.NewServerToClient(
			chain.NewNetworkServiceServer(
				heal.NewServer(ctx, rv),
				clienturl.NewServer(connectTo),
				connect.NewServer(ctx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
					return chain.NewNetworkServiceClient(
						append(
							opts.additionalFunctionality,
							heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
							opts.authorizeClient,
							networkservice.NewNetworkServiceClient(cc),
						)...,
					)
				},
					connect.WithDialOptions(opts.dialOptions...),
					connect.WithDialTimeout(opts.dialTimeout)),
			),
		),
	)

	return *rv
```

**Review by Bolodya1997 — APPROVED** [review]

**Review by edwarnicke — APPROVED** [review]
This LGTM.  Feel free to merge as soon as @Bolodya1997 @denis-tingaikin et al are comfortable with the testing.

**Comment by Bolodya1997** [comment]
Since we already have registry healing merged, lets add with this PR local NSMgr restore test:
```go
func TestNSMGR_HealLocalNSMgrRestored(t *testing.T) {
	nsmgrCtx, nsmgrCtxCancel := context.WithCancel(context.Background())
	defer nsmgrCtxCancel()

	customConfig := []*sandbox.NodeConfig{
		nil,
		{
			NsmgrCtx:               nsmgrCtx,
			NsmgrGenerateTokenFunc: sandbox.GenerateTestToken,
		},
	}

	testNSMGRHealNSMgr(t, 1, customConfig, nsmgrCtxCancel)
}
```
